src/cluster_migrate.c

```diff
@@ -0,0 +1,1986 @@
+                    last_interaction &&
+                    (server.unixtime - last_interaction > server.repl_timeout)) {
+                    serverLog(LL_WARNING,
+                              "Timing out slot migration %s "
+                              "after not receiving ack for too long",
+                              link->description);
+                    finishSlotMigrationLink(link,
+                                            SLOT_MIGRATION_LINK_FAILED,
+                                            "Timed out after too long with no interaction");
+                    continue;
+                }
+                /* Send acks only when the child process isn't writing to it. */
+                if (canSlotMigrationSendAck(link)) {
+                    run_with_period(1000) sendSyncSlotsMessageOnLink(link, "ACK");
+                }
+            }
+        }
+        proceedWithSlotMigration(link);
+    }
+
+    clusterCleanupSlotMigrationLog();
+    slotExportTryUnpause();
+}
+
+void slotExportTryUnpause(void) {
+    if (getPausedActionsWithPurpose(PAUSE_DURING_SLOT_MIGRATION) == 0) return;
+    listNode *ln;
+    listIter li;
+    listRewind(server.cluster->slot_migration_links, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        slotMigrationLink *link = ln->value;
+        if (link->mf_end) {
+            return;
+        }
+    }
+    unpauseActions(PAUSE_DURING_SLOT_MIGRATION);
+}
+
+/* Sent by either the target or the source as a liveness check. */
+void clusterCommandSyncSlotsAck(client *c) {
+    if (!c->slot_migration_link) {
+        serverLog(LL_WARNING,
+                  "Received CLUSTER SYNCSLOTS ACK from client %llu, "
+                  "but the client is not a slot import source or export target. "
+                  "Closing client.",
+                  (unsigned long long)c->id);
+        freeClientAsync(c);
+        return;
+    }
+    slotMigrationLink *link = (slotMigrationLink *)c->slot_migration_link;
+    link->last_ack = server.unixtime;
+    if (link->state == SLOT_IMPORT_WAIT_ACK) {
+        updateSlotMigrationLinkState(link, SLOT_IMPORT_RECEIVE_SNAPSHOT);
+    }
+}
+
+/* Sent by either the target or the source as a control message for progressing
+ * with slot import. */
+void clusterCommandSyncSlots(client *c) {
+    if (c->flag.primary) {
+        /* Due to primary proxying slot migration source commands to replicas, SYNCSLOTS should be
+         * ignored from our primary. */
+        return;
+    }
+    if (!strcasecmp(c->argv[2]->ptr, "establish")) {
+        /* CLUSTER SYNCSLOTS ESTABLISH <args> */
+        clusterCommandSyncSlotsEstablish(c);
+    } else if (!strcasecmp(c->argv[2]->ptr, "snapshot-eof")) {
+        /* CLUSTER SYNCSLOTS SNAPSHOT-EOF */
+        clusterCommandSyncSlotsSnapshotEof(c);
+    } else if (!strcasecmp(c->argv[2]->ptr, "pause")) {
+        /* CLUSTER SYNCSLOTS PAUSE */
+        clusterCommandSyncSlotsPause(c);
```

Comment:
```suggestion
        /* CLUSTER SYNCSLOTS REQUEST-PAUSE */
        clusterCommandSyncSlotsRequestPause(c);
```
Mostly because it's sometimes hard to differentiate pause and paused

Reply:
Sounds good